### PR TITLE
TSFC integration

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR does two things:

- fixes the integration with TSFC and enables expression rewriting
- drastically improves the quality of the codebase

TODO:

- integration of padding + data alignment

Comments welcome 

There are a few pending issues and the need for a little bit of rebasing, but for the rest it should be pretty much ready

### CODE REVIEW

**Comment by dorugeber** [comment]
Ugh.  Obvious comment: the commit is +2400/-2300, and there's lots of 'new files added' and 'old files deleted' which I assume is just code shuffling around.  If you actually want someone to read the PR in detail (maybe you don't!), is there a sane way to split this into a trivial "shuffling" PR and a PR where the changes happen?

**Comment by dorugeber** [comment]
(If it would take you an hour, just do it.  If it would take you a week, don't!)

**Comment by FabioLuporini** [comment]
That's true. It's actually just code refactoring. The only real new stuff is in cse.py, plus a few things in hoister.py. The rest is just code move/renamed/refactored.

Thanks

File: coffee/cse.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
Comment (miklos1) [inline coffee/cse.py]: Wrong year, and possibly wrong copyright holder.

File: coffee/cse.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+
+from plan import verbose
+from base import *
+from utils import *
+from coffee.visitors import EstimateFlops
+from expression import MetaExpr
+
+
+class Temporary():
+
```
Comment (miklos1) [inline coffee/cse.py]: Docstring?

File: coffee/cse.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    def __str__(self):
+        return "%s: level=%d, flops/iter=%d, reads=[%s], isread=[%s]" % \
+            (self.symbol, self.level, self.flops,
+             ", ".join([str(i) for i in self.reads]),
+             ", ".join([str(i) for i in self.is_read]))
+
+
+class CSEUnpicker():
+
```
Comment (miklos1) [inline coffee/cse.py]: Docstring?

File: coffee/cse.py [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+        if verbose:
+            print BLUE % ("Cost model :: unpicking CSE between levels [%d, %d]:" % bounds),
+            print BLUE % ("cost=%d (cse=%d, outloop=%d, inloop_fact=%d, inloop_cse=%d)" %
+                          (uptolevel_cost, cse_cost, total_outloop_cost,
+                           level_inloop_cost, cse))
+
+        return best
+
+    def unpick(self):
+        fors = visit(self.header, info_items=['fors'])['fors']
```
Comment (miklos1) [inline coffee/cse.py]: Docstring?

File: coffee/cse.py [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
Comment (miklos1) [inline coffee/cse.py]: A module docstring would be nice as well.

File: coffee/expander.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
Comment (miklos1) [inline coffee/expander.py]: copy-paste

File: coffee/expander.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        return exp
+
+    def invalidate(self, exp):
+        was_hit = False
+        for i, j in self._map.items():
+            if str(j) == str(exp):
+                self._map.pop(i)
+                if self._hits[i] > 0:
+                    was_hit = True
+        return was_hit
```
Comment (wence-) [inline coffee/expander.py]: AKA:
```python
def invalidate(self, exp):
    for i, j in self._map.items():
        if str(j) = str(exp):
            self._map.pop(i)
            if self._hits[i] > 0:
                return True
     return False
```

File: coffee/rewriter.py [hunk 8]
```diff
@@ -411,593 +446,81 @@ def preevaluate(self):
...
+            # ... define the objective function (min number of factorizations)
+            prob += ilp.lpSum(x[i] for i in nodes_vars)
+
+            # ... solve the problem
+            prob.solve(ilp.GLPK(msg=0))
+
+            # Finally, factorize and hoist (note: the order in which factorizations are carried
+            # out is crucial)
+            nodes = [nodes_vars[n] for n, v in x.items() if v.value() == 1]
+            other_nodes = [nodes_vars[n] for n, v in x.items() if nodes_vars[n] not in nodes]
```
Comment (wence-) [inline coffee/rewriter.py]: If I read the docs for pulp correctly, you're meant to pull the answer and the objective out of the solved problem.  Does it also update variables "in place"?

Comment (wence-) [inline coffee/rewriter.py]: Also, if the order is crucial is it safe to iterate over a dict and collect things?

File: coffee/expander.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+                return (l_exps + r_exps, self.EXPAND)
+            elif l_type == self.EXPAND and r_type == self.EXPAND and isinstance(node, Sub):
+                return (l_exps + [Neg(r) for r in r_exps], self.EXPAND)
+            else:
+                return ([node], self.GROUP)
+
+        else:
+            raise RuntimeError("Expansion error: unknown node: %s" % str(node))
+
+    def expand(self, should_expand, **kwargs):
```
Comment (miklos1) [inline coffee/expander.py]: docstring?

File: coffee/expression.py [hunk 10]
```diff
@@ -155,16 +172,11 @@ def mode(self, value):
...
-    ``kwargs``. In particular, ``kwargs`` has the following keys:
-
-        * parent: the block node that embeds the expression.
-        * loops_info: an iterator of 2-tuple ``(loop, loop_parent)`` which
-          substitute analogous information in the new ``MetaExpr``.
-        * domain: the domain of the output tensor evaluated by the expression.
-        * mode: the suggested rewrite mode.
-    """
+    ``kwargs``. ``kwargs`` accepts the following keys: parent, loops_info,
+    linear_dims, mode."""
```
Comment (miklos1) [inline coffee/expression.py]: Why did you remove the explanation of `parent`, `loops_info` etc. ?

File: coffee/factorizer.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
Comment (miklos1) [inline coffee/factorizer.py]: copy-pasta

File: coffee/factorizer.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+    @staticmethod
+    def process(symbols, should_factorize, op=None):
+        operands = [s for s in symbols if should_factorize(s)]
+        factors = [s for s in symbols if not should_factorize(s)]
+        return Term(operands, factors, op)
+
+
+class Factorizer():
+
```
Comment (miklos1) [inline coffee/factorizer.py]: docstring?

File: coffee/factorizer.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+                    _t.add_factors(factor)
+            factorized = [t.generate_ast for t in factorized.values()]
+            factorized = ast_make_expr(Sum, factorized)
+            parent.children[parent.children.index(node)] = factorized
+            return Term([], [factorized])
+
+        else:
+            return Term([], [node])
+
+    def factorize(self, should_factorize, **kwargs):
```
Comment (miklos1) [inline coffee/factorizer.py]: docstring?

File: coffee/hoister.py [hunk 14]
```diff
@@ -0,0 +1,431 @@
+# This file is part of COFFEE
+#
+# COFFEE is Copyright (c) 2014, Imperial College London.
```
Comment (miklos1) [inline coffee/hoister.py]: copy-pasta

**Comment by miklos1** [comment]
This patch removes 22 `dcopy`s and adds 13. I guess that is an improvement.

File: coffee/rewriter.py (hunk 8 above)
Comment (FabioLuporini) [inline coffee/rewriter.py]: Yes. I found it in some tutorials tbh

As for the order: what I actually meant is that all ``nodes`` have to come before all ``other_nodes``

File: coffee/expander.py (hunk 9 above)
Comment (FabioLuporini) [inline coffee/expander.py]: just the main expander function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

File: coffee/expander.py (hunk 7 above)
Comment (FabioLuporini) [inline coffee/expander.py]: true :) changed

File: coffee/expander.py (hunk 6 above)
Comment (FabioLuporini) [inline coffee/expander.py]: fixed

File: coffee/cse.py (hunk 4 above)
Comment (FabioLuporini) [inline coffee/cse.py]: just the main cseunpicker function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

File: coffee/cse.py (hunk 3 above)
Comment (FabioLuporini) [inline coffee/cse.py]: added

File: coffee/cse.py (hunk 2 above)
Comment (FabioLuporini) [inline coffee/cse.py]: added

File: coffee/cse.py (hunk 1 above)
Comment (FabioLuporini) [inline coffee/cse.py]: fixed the year

File: coffee/expression.py (hunk 10 above)
Comment (FabioLuporini) [inline coffee/expression.py]: it was basically the same description as the one you can find in the MetaExpr class' docstring

File: coffee/factorizer.py (hunk 11 above)
Comment (FabioLuporini) [inline coffee/factorizer.py]: fixed

File: coffee/factorizer.py (hunk 12 above)
Comment (FabioLuporini) [inline coffee/factorizer.py]: just the main factorizer function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

File: coffee/factorizer.py (hunk 13 above)
Comment (FabioLuporini) [inline coffee/factorizer.py]: just the main factorizer function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)

**Comment by FabioLuporini** [comment]
Addressed most of your comments

Please pull

Now the zero-elimination engine is integrated. As a temporary experiment, I've added this transformation to the default optimisation pipeline (i.e., O2) . @wence-  @miklos1  Can you please beta-test this branch at least for functionality ? I'm still running a few experiments for performance. 

Thanks

File: coffee/factorizer.py (hunk 12 above)
Comment (miklos1) [inline coffee/factorizer.py]: meh... how am I supposed to know that? Maybe add some reference at least.

**Comment by miklos1** [comment]
Did you just fix -O4? The hyperelasticity test is unexpectedly passing again.

**Comment by wence-** [comment]
So that xfail was on old pre-tsfc coffee.  I can entirely imagine that many things have changed.  For example, if no optimisations occur then the test xpasses.

**Comment by miklos1** [comment]
With last week's `tsfc-updates` COFFEE, it was still xfailing, but now it passes, and COFFEE claims some flop reduction (though not as much as with -O2).

**Comment by FabioLuporini** [comment]
I've been changing the transformations applied by each optimisation level. 

Now -O4 does what previously -O3 did. Some fixes went it and made it work. 

Also (see last commit in capital letters), at the moment O2 == O3. I've temporarily disabled alignment and padding in {O2,O3} because most users use gcc, and gcc doesn't like auto-vectorising the code, despite all my efforts trying to generate "easily vectorisable source code". Alignment and padding are applied instead at level O4.

I was thinking that one way of deciding whether to apply padding and data alignment is to check whether the underlying compiler is "recent enough" to support the #pragma simd directive (which enforces vectorisation). Most recent versions of gcc have it. We could think of spoofing somehow the compiler version...

File: coffee/factorizer.py (hunk 13 above)
Comment (FabioLuporini) [inline coffee/factorizer.py]: ok done now

File: coffee/factorizer.py (hunk 12 above)
Comment (FabioLuporini) [inline coffee/factorizer.py]: ok doing it now

**Comment by FabioLuporini** [comment]
Added logic for avoiding zero elimination 

Addressed your last comments concerning docstrings

Tests are now running. Maybe it is also now a good time to purge COFFEE from crap ``Par`` nodes :) a git grep seems to show they are never used to construct objects...

File: .travis.yml [hunk 15]
```diff
@@ -10,6 +10,11 @@ language: python
 python:
   - "2.7"
 
+addons:
+  apt:
+    packages:
+    - glpk-utils
```
Comment (kynan) [inline .travis.yml]: Is this going to prevent Travis from running on container infrastructure?

File: coffee/base.py [hunk 16]
```diff
@@ -596,8 +601,13 @@ def rvalue(self, val):
         self.children[1] = val
 
     def gencode(self, not_scope=False):
```
Comment (kynan) [inline coffee/base.py]: Double negation? `not_scope=False` == `scope=True`?

File: coffee/cse.py [hunk 17]
```diff
@@ -0,0 +1,419 @@
...
+import operator
+
+from plan import verbose
+from base import *
+from utils import *
+from coffee.visitors import EstimateFlops
+from expression import MetaExpr
+
+
+class Temporary():
```
Comment (kynan) [inline coffee/cse.py]: New style class?

File: .travis.yml (hunk 15 above)
Comment (FabioLuporini) [inline .travis.yml]: I honestly have no idea... @miklos1  ?

File: coffee/base.py (hunk 16 above)
Comment (FabioLuporini) [inline coffee/base.py]: yes that's bad... I should maybe change it

File: coffee/cse.py (hunk 17 above)
Comment (FabioLuporini) [inline coffee/cse.py]: do you have a preference ?

File: .travis.yml (hunk 15 above)
Comment (miklos1) [inline .travis.yml]: I mostly added this as documentation, because COFFEE now relies on this, even though its test suite does not actually uses it. I have no idea about Travis.

File: coffee/cse.py (hunk 17 above)
Comment (miklos1) [inline coffee/cse.py]: New style class is always better.

**Comment by miklos1** [comment]
Purged `Par` nodes, so this pull request now depends on firedrakeproject/firedrake#746.

File: coffee/cse.py [hunk 18]
```diff
@@ -0,0 +1,419 @@
...
+        cost = 0
+        for level, temporaries in levels.items():
+            cost += sum(t.flops*t.niters for t in temporaries)
+        return cost
+
+    def _cost_fact(self, trace, levels, bounds):
+        # Check parameters
+        bounds = bounds or (min(levels.keys()), max(levels.keys()))
+        assert len(bounds) == 2 and bounds[1] >= bounds[0]
+        assert [i in levels.keys() for i in bounds]
```
Comment (kynan) [inline coffee/cse.py]: This does't do what you expect it to

Comment (wence-) [inline coffee/cse.py]: By which he means you wanted `for i in bounds: assert i in levels.keys()`

Comment (wence-) [inline coffee/cse.py]: Better to write: `assert set(levels.keys()) == set(bounds)` I think.

File: coffee/cse.py [hunk 19]
```diff
@@ -0,0 +1,419 @@
...
+        fact_levels = OrderedDict([(k, v) for k, v in levels.items()
+                                   if k > bounds[0] and k <= bounds[1]])
```
Comment (wence-) [inline coffee/cse.py]: It's unclear why you would want an ordered dict here given that levels is an unordered dict.

File: coffee/cse.py (hunk 17 above)
Comment (FabioLuporini) [inline coffee/cse.py]: OK

File: coffee/cse.py (hunk 18 above)
Comment (FabioLuporini) [inline coffee/cse.py]: uooops! Fixing

File: coffee/cse.py (hunk 19 above)
Comment (FabioLuporini) [inline coffee/cse.py]: good spot

**Comment by FabioLuporini** [comment]
Addressed latest comments (tests running now)

File: coffee/expander.py [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+        no new symbols need be introduced. Otherwise (e.g., the H temporary
+        appears in multiple places), create a new symbol."""
+        exp, grp = expansion.left, expansion.right
+
+        # First, check if any of the symbols in /exp/ have been hoisted
+        try:
+            retval = FindInstances.default_retval()
+            exp = [s for s in FindInstances(Symbol).visit(exp, ret=retval)[Symbol]
+                   if s.symbol in self.hoisted and self.should_expand(s)][0]
+        except:
```
Comment (kynan) [inline coffee/expander.py]: How would the above fail?

Comment (FabioLuporini) [inline coffee/expander.py]: [0] at the end

Comment (miklos1) [inline coffee/expander.py]: This is very loose, because `except:` will catch expression thrown from anywhere inside this expression. Can you make it more specific?

Comment (wence-) [inline coffee/expander.py]: So you're checking if len(...) == 0?

Comment (FabioLuporini) [inline coffee/expander.py]: yes. OK, I'll make it more specific

File: coffee/hoister.py [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from warnings import warn as warning
+
+from base import *
+from utils import *
+
+
+class Extractor():
```
Comment (kynan) [inline coffee/hoister.py]: New style class?

File: coffee/factorizer.py [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+                self.factors.remove(f)
+
+    @staticmethod
+    def process(symbols, should_factorize, op=None):
+        operands = [s for s in symbols if should_factorize(s)]
+        factors = [s for s in symbols if not should_factorize(s)]
+        return Term(operands, factors, op)
+
+
+class Factorizer():
```
Comment (kynan) [inline coffee/factorizer.py]: New style class?

File: coffee/factorizer.py [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import operator
+
+from base import *
+from utils import *
+
+
+class Term():
```
Comment (kynan) [inline coffee/factorizer.py]: New style class?

File: coffee/hoister.py [hunk 24]
```diff
@@ -0,0 +1,431 @@
...
+                # E.g. A[i]*B[j], hoistable in TMP[i,j]
+                return (dep_n, self.EXT)
+        elif info_r == self.EXT:
+            self._try(right, dep_r)
+        elif info_l == self.EXT:
+            self._try(left, dep_l)
+        return (dep_n, self.STOP)
+
+
+class Hoister():
```
Comment (kynan) [inline coffee/hoister.py]: New style class?

File: coffee/hoister.py [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+
+    # How many times the hoister was invoked
+    _handled = 0
```
Comment (kynan) [inline coffee/hoister.py]: Suppose this functionality could be factored into a common base class?

File: coffee/rewriter.py [hunk 26]
```diff
@@ -31,18 +31,20 @@
...
-from ast_analyzer import ExpressionGraph, StmtTracker
 from coffee.visitors import *
+from hoister import Hoister
+from expander import Expander
+from factorizer import Factorizer
+from cse import CSEUnpicker
 
 
-class ExpressionRewriter(object):
+class ExpressionRewriter():
```
Comment (kynan) [inline coffee/rewriter.py]: Why make this an old style class?

File: coffee/utils.py [hunk 27]
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
         return loops
 
 
+###############################################################
+# Utilities for tracking the global impact of transformations #
+###############################################################
+
+
+class StmtTracker(OrderedDict):
```
Comment (kynan) [inline coffee/utils.py]: I suppose a `StmtTracker` HAS-A `OrderedDict` rather than IS-A `OrderedDict`?

File: coffee/utils.py [hunk 28]
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
...
+
+    whose semantics is, respectively, as follows:
+
+        * The AST node whose ``str(lvalue)`` is used as dictionary key
+        * The AST node of the symbol declaration
+        * The AST node of the closest loop enclosing the statement
+        * The parent of the closest loop
+    """
+
+    class StmtInfo(object):
```
Comment (kynan) [inline coffee/utils.py]: Use a [`namedtuple`](https://docs.python.org/2/library/collections.html#collections.namedtuple) instead?

File: requirements.txt [hunk 29]
```diff
@@ -1 +1,2 @@
 networkx
+PuLP
```
Comment (kynan) [inline requirements.txt]: Does COFFEE work with the latest PuLP release?

Comment (FabioLuporini) [inline requirements.txt]: The latest is 1.6.1 and it does

File: coffee/utils.py (hunk 28 above)
Comment (FabioLuporini) [inline coffee/utils.py]: Ah, I didn't know about it

File: coffee/utils.py (hunk 27 above)
Comment (FabioLuporini) [inline coffee/utils.py]: IS-A

Comment (wence-) [inline coffee/utils.py]: Confusingly, it both *has* an ordereddict and *is* an ordereddict :).

File: coffee/utils.py [hunk 30]
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
...
+                assert(k in self.__class__.INFO)
+                setattr(self, k, v)
+
+    def __init__(self):
+        super(StmtTracker, self).__init__()
+        self.byvalue = OrderedDict()
+
+    def __setitem__(self, key, value):
+        if not isinstance(value, self.StmtInfo):
+            if not isinstance(value, tuple):
```
Comment (wence-) [inline coffee/utils.py]: `if not isinstance(value, (tuple, self.StmtInfo))`

Comment (wence-) [inline coffee/utils.py]: Oh, I see why not.

File: coffee/utils.py [hunk 31]
```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
...
+                raise RuntimeError("StmtTracker accepts tuple or StmtInfo objects")
+            value = self.StmtInfo(**dict(zip(self.StmtInfo.INFO, value)))
```
Comment (wence-) [inline coffee/utils.py]: `zip` throws away entries from the longer list.  Do you want to assert that `len(value) == len(self.StmtInfo.INFO`?

File: coffee/utils.py (hunk 27 above)
Comment (FabioLuporini) [inline coffee/utils.py]: Ah right I had forgotten about it. It's because I need access by value too at some point

File: coffee/utils.py (hunk 31 above)
Comment (FabioLuporini) [inline coffee/utils.py]: adding the check

File: coffee/expander.py (hunk 20 above)
Comment (FabioLuporini) [inline coffee/expander.py]: I did better than that, I deleted a whole chunk of code that was only useful in a negligible number of cases

**Comment by FabioLuporini** [comment]
OK, again, addressed your comments.

**Comment by miklos1** [comment]
So it seems that COFFEE only works in addition to the "unroll `IndexSum`s" TSFC optimisation. If that optimisation is disabled (it is on by default), then COFFEE fails. This is not blocker, but it is nice to have it documented.

**Comment by wence-** [comment]
fails, or does nothing?

**Comment by miklos1** [comment]
Fails.

**Comment by miklos1** [comment]
firedrakeproject/firedrake#746 is merged. @FabioLuporini: get this ready for the merge.
